### PR TITLE
Export current criu mode to opts

### PR DESCRIPTION
**Pull request by Snorch**
We have multiple options which are valid only on restore or only on dump or in any other specific criu mode, so it would be useful to have info about current mode in opts so that we can validate other options against current mode.

Plan is to use it for mount-v2 (switch from current restore mount engine to a new one) option as it is only valid on restore, and this would make handling of different types of mountpoints much easier.

- When criu mode is set from main() we just parse mode from argv[optind] just after parse_options() found optind of the command. Note that opts.mode is available before check_options().

- For rpc service we reset opts.mode to CR_SWRK each time we restart cr_service_work(), in the original service process we still have CR_SERVICE to differentiate between them, and each request handling function which does setup_opts_from_req sets opts.mode in accordance with the processed request type. And it is also available before check_options().
    
Now in check_options we can add filters on one mode only options.

Signed-off-by: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>

Note: I chose to call this "mode" as in rpc we have a request type and on cli we have command, so "mode" should unify these two similar terms in one.

### CODE REVIEW

**Comment by Snorch** [comment]
@checkpoint-restore/maintainers Please take a look, hope this would simplify everything. I thought about a way to differentiate between dump/restore already for a long time, in many places we already pass restore/dump info in function parameters, maybe switching to this approach would be better.

**Review by adrianreber — APPROVED** [review]
Sounds like a good idea and all CI tests are happy. :+1:

**Review by mihalicyn — APPROVED** [review]

File: criu/include/cr_options.h [hunk 1]
```diff
@@ -100,6 +100,22 @@ struct irmap_path_opt {
...
+	CR_LAZY_PAGES,
+	CR_CHECK,
+	CR_PAGE_SERVER,
+	CR_SERVICE,
+	CR_SWRK,
+	CR_DEDUP,
+	CR_CPUINFO,
+	CR_EXEC_DEPRECATED,
+	CR_SHOW_DEPRECATED,
+};
```
Comment (rst0git) [inline criu/include/cr_options.h:117]: Would it be possible to update `image_dir_mode()` in `crtools.c` with these enum values?

Comment (Snorch) [reply criu/include/cr_options.h:117]: Thanks for review! Sure, I forgot about it.

File: criu/crtools.c [hunk 2]
```diff
@@ -56,13 +56,20 @@ void flush_early_log_to_stderr(void)
...
 		return O_RSTR;
+	case CR_CPUINFO:
+		if (!strcmp(argv[optind + 1], "dump"))
+			return O_DUMP;
+		/* fallthrough */
+	default:
+		return -1;
+	}
 
 	return -1;
```
Comment (rst0git) [inline criu/crtools.c:74]: This `return -1;` is not used.

File: criu/crtools.c [hunk 3]
```diff
@@ -185,7 +222,7 @@ int main(int argc, char *argv[], char *envp[])
 	}
 
 	/* We must not open imgs dir, if service is called */
-	if (strcmp(argv[optind], "service")) {
+	if (opts.mode != CR_SERVICE) {
```
Comment (rst0git) [inline criu/crtools.c:232]: Could you please add something like the following to fix a segmentation fault for `criu cpuinfo`?
```C
if (opts.mode == CR_CPUINFO && argc < 3) {
        pr_err("cpuinfo requires an action: dump or check\n");
        goto usage;
}
```

File: criu/crtools.c (hunk 2 above)
Comment (Snorch) [reply criu/crtools.c:74]: Sure I know, it is a combination of two best practices: 1) switch-case should have explicit default 2) int function should have proper return at the end.

I can add a comment /* Should never get here */ if needed.

File: criu/crtools.c (hunk 3 above)
Comment (Snorch) [reply criu/crtools.c:232]: Ok, will do.

**Comment by adrianreber** [comment]
Yes, @rst0git also mentioned it. Very strange, not sure what it is.

@rst0git any insights yet?

It is a bit worrying that so many tests are failing with a segfault.

**Pull request merged**